group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2049,6 +2099,11 @@ private CoordinatorResult<Void, CoordinatorRecord> classicGroupJoinToConsumerGro
                 scheduleConsumerGroupSyncTimeout(groupId, response.memberId(), request.rebalanceTimeoutMs());
 
                 responseFuture.complete(response);
+
+                // Maybe downgrade the consumer group if the last member using the
+                // consumer protocol is replaced by the joining member. No rebalance
+                // is needed after the replacement.
+                scheduleConsumerGroupDowngradeTimeout(group, false);
```

Comment:
I wonder if we should rather call this one out side of the `appendFuture`. My concern is that we call `validateOnlineDowngrade` in `scheduleConsumerGroupDowngradeTimeout` and it accesses the uncommitted state of the group while we actually complete the write operation. This may lead to unexpected results.